valhalla/jawn/src/managers/traceManager.ts

```diff
@@ -0,0 +1,315 @@
+          index: number,
+          logprobs: null,
+          finish_reason: string,
+          message: {
+            role: "system" | "user" | "assistant",
+            content: string,
+          }
+        }[]);
+  };
+
+  private async storeRawLogsS3(span: TModifiedSpan, requestBody: TRequestBody, responseBody: TResponseBody, authParams: AuthParams) {
+    const key = this.s3Client.getRawRequestResponseKey(span.traceId, authParams.organizationId);
+    const s3Result = await this.s3Client.store(
+      key,
+      JSON.stringify({
+        request: JSON.stringify(requestBody),
+        response: JSON.stringify(responseBody)
+      }),
+    );
+    if (s3Result.error) {
+      console.error(`Error storing request response in S3: ${s3Result.error}`);
+    }
+  }
+
+  private constructLog(
+    span: TModifiedSpan,
+    promptMessages: {
+      role: "system" | "user" | "assistant";
+      content: string;
+    }[],
+    completionChoices: TCompletionChoices,
+    authParams: AuthParams
+  ): Log {
+    return {
+      request: {
+        id: span.traceId,
+        userId: authParams.userId ?? "",
+        promptId: undefined,
+        properties: {},
+        heliconeApiKeyId: authParams.heliconeApiKeyId ?? undefined,
+        heliconeProxyKeyId: undefined,
+        targetUrl: "",
+        provider: span.attributes.get("gen_ai.system"),
+        bodySize: JSON.stringify(promptMessages).length,
+        path: "async-unknown-path",
+        threat: false,
+        countryCode: undefined,
+        requestCreatedAt: new Date(parseInt(span.startTimeUnixNano) / 1000000),
+        isStream: false,
+        heliconeTemplate: undefined,
+      },
+      response: {
+        id: span.traceId,
+        status: 200,
+        bodySize: JSON.stringify(completionChoices).length,
+        timeToFirstToken: undefined,
+        responseCreatedAt: new Date(parseInt(span.endTimeUnixNano) / 1000000),
+        delayMs: -1,
+      }
+    };
+  }
+
+  private async sendLogToKafka(kafkaMessage: Message) {
+    const kafkaProducer = new KafkaProducer();
+    await kafkaProducer.sendMessages(
+      [kafkaMessage],
+      "request-response-logs-prod"
+    );
+  }
+
+  private processOtelSpans(trace: OTELTrace): TModifiedSpan[] {
+    return trace.resourceSpans.flatMap(resourceSpan =>
+      resourceSpan.scopeSpans.flatMap(scopeSpan =>
+        scopeSpan.spans.map(span => {
+          const attributes: Map<string, any> = new Map();
+          span.attributes.forEach(({ key, value }) => {
+            attributes.set(key, value.stringValue ?? value.intValue);
+          });
+
+          const uuid = randomUUID();
+          return {
+            ...span,
+            traceId: uuid,
+            attributes: attributes,
+          };
+        })
+      )
+    );
+  }
+
+  public async consumeTraces(trace: OTELTrace, heliconeAuthorization: string, authParams: AuthParams) {
+    const spans = this.processOtelSpans(trace);
+
+    spans.forEach(async span => {
```

Comment:
This function is not garaunteed to end when `consumeTraces` ends. 

Please replace this with 

```suggestion
    for(const span in spans) {
```

This will ensure that anything awaited in the body is awaited before `consumeTraces` returns